remote_execution/oss/re_grpc/src/client.rs

```diff
@@ -466,77 +577,240 @@ impl REClient {
+            let metadata = metadata.clone();
+            let fut = async move {
+                // Number of complete (non-partial) messages
+                let num_msgs = size as usize / max_msg_size;
+                let upload_segments = futures::stream::iter(data).chunks(max_msg_size);
+                let requests = upload_segments
+                    .enumerate()
+                    .map(move |(i, chunk)| WriteRequest {
+                        resource_name: resource_name.to_owned(),
+                        write_offset: (i * max_msg_size) as i64,
+                        finish_write: i == num_msgs,
+                        data: chunk.to_owned(),
+                    });
+
+                let resp = bytestream_client
+                    .write(with_internal_metadata(requests, metadata))
+                    .await?;
+                if resp.into_inner().committed_size != size as i64 {
+                    return Err(anyhow::anyhow!(
+                        "Failed to upload inline blob: invalid committed_size from WriteResponse"
+                    ));
+                }
+
+                Ok(vec![hash])
+            };
+            upload_futures.push(Box::pin(fut));
+        }
+
+        // Create futures for any files that needs uploading.
+        for file in request.files_with_digest.unwrap_or_default() {
+            let mut bytestream_client = self.grpc_clients.bytestream_client.clone();
+            if !needed_blobs.contains(&file.digest) {
+                // Blob is already in remote, skip upload.
+                continue;
+            }
+
+            let hash = file.digest.hash.clone();
+            let size = file.digest.size_in_bytes;
+            let name = file.name.clone();
+            let metadata = metadata.clone();
+            if size < max_msg_size as i64 {
+                batched_blob_updates.push(BatchUploadRequest::File(file));
+                continue;
+            }
+            let resource_name = format!("uploads/{}/blobs/{}/{}", client_uuid, hash.clone(), size);
+            let fut = async move {
+                let file = tokio::fs::File::open(&name).await?;
+                let mut reader = BufReader::with_capacity(max_msg_size, file);
+                let mut write_offset = 0;
+                let mut upload_segments = Vec::new();
+                loop {
+                    let data = reader.fill_buf().await?;
+                    let length = data.len();
+
+                    upload_segments.push(WriteRequest {
+                        resource_name: resource_name.to_owned(),
+                        write_offset,
+                        finish_write: length == 0,
+                        data: data.to_owned(),
+                    });
+
+                    write_offset += length as i64;
+                    reader.consume(length);
+                    if length == 0 {
+                        break;
```

Comment:
Is the use of a buffered reader all that useful here? This adds the complexity of lining up the `consume` calls but without real benefit. We are just going to issue reads for `max_msg_size` anyway:

```suggestion
            let fut = async move {
                let file = tokio::fs::File::open(&name).await?;
                let mut buff = vec![0; max_msg_size];
                
                let mut write_offset = 0;
                let mut upload_segments = Vec::new();
                
                loop {
                    let length = file.read(&mut buff).await.context("Error reading")?;

                    upload_segments.push(WriteRequest {
                        resource_name: resource_name.to_owned(),
                        write_offset,
                        finish_write: length == 0,
                        data: data.to_owned(),
                    });

                    write_offset += length as i64;

                    if length == 0 {
                        break;
```